### ISSUE TITLE
Introduce alternative names for C types of specific bitwidths

### ISSUE THREAD

**Issue by nomennescio** [factor/factor#2612]
At many places in Factor's vocabularies, alien C types are used. Unfortunately, C types bit-widths are unspecified, whereas Factor dictates a certain bit-width for these types, suggesting a uniform approach when translating source code where it doesn't exist, as it must be checked that these assumptions are valid for a specific compiler and platform.

As an alternative, C has always had types to unambiguously use types of certain bit-widths, in `<stdint.h>`:
```
int8_t
int16_t
int32_t
uint8_t
uint16_t
uint32_t
int64_t
uint64_t
```

I propose to add these type names as alternative names where possible.

Next to that, I would like to have yet another set of alternative names, which are more brief, and which are also more suitable to be used in `math.vectors.simd` types and the likes, these are
```
i8
i16
i32
u8
u16
u32
i64
u64
```

The corresponding 128 bit SIMD types could the be aliased like:
```
i8-16
i16-8
i32-4
i64-2
u8-16
u16-8
u32-4
u64-2
```

For floating point a similar naming scheme could be used then:

```
f32
f64
```
and SIMD types
```
f32-4
f64-2
```

As these third set of types names are almost the same length, this will lead to easier to read code, and I would prefer to use this third set as the default names for future vocabularies using C typenames for fixed bit-width types.

I think this would also lead to nicer formatted and less ambiguous code. It also can be easily extended for future `simd-256` and `simd-512` types.

**Comment by mrjbq7** [comment]
Don't we have those?

https://docs.factorcode.org/content/vocab-alien.c-types.html

Is your suggestion to use them in more places?

I'm not sure that all C API's support these versus the platform-specific versions (e.g., char, int, long, longlong, etc).

**Comment by nomennescio** [comment]
My bad, I should have checked with the latest documentation. I'm amazed even the abbreviated types are there, with the exact same names as I've proposed. But indeed, the major suggestion is to use them at more places, because when looking at "alien" code, I noticed the use of char, int, longlong etc. almost everywhere, which triggered me to submit this issue, assuming these types were not available at all.

As for the SIMD ones, these seem like a more natural way to describe these types. Even the Intel/Microsoft way to describe vector simd ops with WORD, DWORD, QWORD etc. is bad, although it's specified. 

My proposal for the vector types neatly aligns with the abbreviated C types, and all have the same string length as a bonus.

This all seems more like a "convention" issue, but I think it would improve code and documentation readability and comprehensibility. I could create a pull request out of it, if we can agree to what extend you would want to add this to the existing code base.

**Comment by nomennescio** [comment]
> I'm not sure that all C API's support these versus the platform-specific versions (e.g., char, int, long, longlong, etc).

Can you explain the issue you see?

**Comment by nomennescio** [comment]
> (I mean, not the SIMD ones, those look new)

I'm not sure if using the dash in the name, not an underscore, would work in all usecases, but it's using the same convention as the current Factor types. 

I've seen some SIMD naming schemes where they use a letter to indicate the multiplicity, or don't indicate multiplicity at all, but only the subtype lengths, and either implicitly or explicitly indicate the length of the total vector (e.g by using the X, Y, Z vector register name prefixes). I'm not sure if there's an alternative scheme I would prefer, but it should be concise and unambiguous, and preferably of (near) equal string length.

**Comment by melted** [comment]
I don't see how we can use more precise types if the C libraries don't. If the C header uses `int` and `long` and similar types in its prototypes, and a lot of them do, we would have to do the same to be compatible.

**Comment by nomennescio** [comment]
The thing is, Factor assumes certain bitwidths already, so it only gives the illusion of compatibility. And next to that, typedefs are irrelevant when linking to object code using `alien` anyway, only bitwidth and signed-ness.

**Comment by mrjbq7** [comment]
Factor endeavors to make “int” match the platforms definition which on one might be int32 and on another might be int16.

**Comment by mrjbq7** [comment]
If your C api uses int32_t then that’s what Factor should use, but if it uses int, then that’s also what Factor should use.

**Comment by nomennescio** [comment]
> Factor endeavors to make “int” match the platforms definition which on one might be int32 and on another might be int16.
> […](#)
> On Jun 17, 2022, at 11:11 AM, nomennescio ***@***.***> wrote: ﻿ The thing is, Factor assumes certain bitwidths already, so it only gives the illusion of compatibility. And next to that, typedefs are irrelevant when linking to object code using alien anyway, only bitwidth and signed-ness. — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you commented.



> If your C api uses int32_t then that’s what Factor should use, but if it uses int, then that’s also what Factor should use.
> […](#)
> On Jun 17, 2022, at 11:11 AM, nomennescio ***@***.***> wrote: ﻿ The thing is, Factor assumes certain bitwidths already, so it only gives the illusion of compatibility. And next to that, typedefs are irrelevant when linking to object code using alien anyway, only bitwidth and signed-ness. — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you commented.

I think the issue is here with "*match* the *platform* definitions" and "what Factor *should* use". First of all, Factor already defines the bitwidths for C types, so I fail to see how the matching would work. Secondly, "the platform", is a bit too generic, as there might be different compilers involved (e.g. existing binaries). I think this boils down to how Factor *currently* handles its own platform dependence, which has certainly room for improvement, as it now "locks it in" to the platform it is compiled on. I'm used to developing cross-platform, or even cross-compiling or cross-linking, where native types and compiler versions should not matter, but binary compatibility does. So the Factor approach looks too restrictive to me, based on too many assumptions. It might be convenient for some use cases, but limits other use cases, such as being binary compatible. 

I recognize this is  a broader scope, and I'm not sure if this CR is the correct place to discuss this.

To summarize; Factor already assumes sizes of all C integer types, why not represent that more in the Factor code?

**Comment by nomennescio** [comment]
https://docs.factorcode.org/content/article-c-types.primitives.html

```

char | always 1 byte
uchar |  
short | always 2 bytes
ushort |  
int | always 4 bytes
long | same size as CPU word size and void*, except on 64-bit Windows, where it is 4 bytes
longlong | always 8 bytes
```

**Comment by nomennescio** [comment]
(To be clear, there's only exceptional situations where it's not safe to assume the width of C's char to be 8 bits. Other types' sizes are always multiples of char's bit size, because char is the smallest addressable unit of memory in C)

**Comment by nomennescio** [comment]
An example from @mrjbq7 recent presentation:
```
HOOK: rdtsc cpu ( -- n )
M: x86.64 rdtsc
    longlong { } cdecl [
        RAX 0 MOV
        RDTSC
        RDX 32 SHL
        RAX RDX OR
    ] alien-assembly ;
```

Here `longlong` is a 64 bit integer, which is not specified by the C standard. To me it would be better to use `int64_t` or `i64`.
